### ISSUE TITLE
Isn't the softmax of the attention matrix calculated incorrectly?

### ISSUE THREAD

**Issue by PascalHbr** [lucidrains/vit-pytorch#44]
The softmax is currently calculated as follows:

`attn = dots.softmax(dim=-1) `

The softmax is only applied on the last dimension instead of the whole matrix. Instead, I think it should be implemented like that:

`attn = rearrange(dots, 'b h i j -> b h (i j)').softmax(dim=-1).view_as(dots)`

Is this correct or am I missing anything?

**Comment by lucidrains** [comment]
@PascalHbr Hello! So I believe it is correct the way it is

The image has already been cut into patches and linearized into tokens. The dimensions of `i` and `j` are actually the similarity of every patch to every other patch including itself. You then force the network to pick (with softmax) the patch that should be attended to. This is done on the `j` dimension, so that each patch `i` picks one of the other patches that is most important.

Perhaps you were thinking there is some correspondence between `i`, `j` and the height and width, but `i` and `j` are actually height * width.

**Comment by PascalHbr** [comment]
Thanks a lot for the clarification. You are right indeed :)
